File: scala-client/src/main/java/tech/mlsql/byzer_client_sdk/scala_lang/generator/builder.scala
package tech.mlsql.byzer_client_sdk.scala_lang.generator

import net.sf.json.JSONObject

import java.util.UUID
import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

trait BaseNode {
  def tableName: String

  def namedTableName(tableName: String): BaseNode

  def tag(str: String): BaseNode

  def end: Byzer

  def options(): Options

  def toBlock: String

  def getTag: Option[String]

  def toJson: String

  def fromJson(json: String): BaseNode
}


case class OptionValue(value: String, quoteStr: Option[String])

class Options(parent: BaseNode) {
  private val o = new mutable.HashMap[String, OptionValue]()

  def add(name: String, value: String): Options = {
    o.put(name, OptionValue(value, None))
    this
  }

  def addWithQuotedStr(name: String, value: OptionValue) = {
    o.put(name, value)
    this
  }

  def end = {
    parent
  }

  def items = o.toMap

  def toFragment = {
    val opts = o.map { case (k, v) =>
      if (v.quoteStr.isEmpty) {
        if (v.value.isEmpty) {
          s"""`${k}`="""""
        } else {
          s"""`${k}`='''${v.value}'''"""
        }
      } else {
        s"""`${k}`=${v.quoteStr}${v}${v.quoteStr}"""
      }

    }.mkString(" and ")

    val optStr = if (!o.isEmpty) {
      s"""where ${opts}"""
    } else {
      ""
    }
    optStr
  }
}

case class MetaMeta(name: String)

trait FilterNode {

  def toFragment = {
    visit(this)
  }

  def visit(node: FilterNode): String = {
    if (node.isInstanceOf[And]) {
      val n = node.asInstanceOf[And]
      s"""(${visit(n.left)} and ${visit(n.right)})"""
    } else if (node.isInstanceOf[Or]) {
      val n = node.asInstanceOf[Or]
      s"""(${visit(n.left)} or ${visit(n.right)})"""
    } else {
      node.asInstanceOf[Expr].expr.get
    }
  }
}

trait FilterNodeMeta {}

case class AndFilterNodeMeta(k: String, left: FilterNodeMeta, right: FilterNodeMeta) extends FilterNodeMeta

case class OrFilterNodeMeta(k: String, left: FilterNodeMeta, right: FilterNodeMeta) extends FilterNodeMeta

case class ExprFilterNodeMeta(k: String, v: Expr) extends FilterNodeMeta

case class And(left: FilterNode, right: FilterNode) extends FilterNode

case class Or(left: FilterNode, right: FilterNode) extends FilterNode

case class Expr(expr: Option[String]) extends FilterNode

trait BaseOpt {
  def toFilterNode: FilterNode
}

object AndOrTools {
  def toFilterNode(json: JSONObject): FilterNode = {

    json.getString("k") match {
      case "or" =>
        val left = json.getJSONObject("left")
        val right = json.getJSONObject("right")
        Or(toFilterNode(left), toFilterNode(right))
      case "and" =>
        val left = json.getJSONObject("left")
        val right = json.getJSONObject("right")
        And(toFilterNode(left), toFilterNode(right))
      case "expr" =>
        val v = json.getJSONObject("v")
        Expr(Some(v.getString("expr")))
    }
  }

  def toFilterNodeMeta(v: FilterNode): FilterNodeMeta = {
    v match {
      case _ if v.isInstanceOf[And] =>
        val vv = v.asInstanceOf[And]
        AndFilterNodeMeta("and", toFilterNodeMeta(vv.left), toFilterNodeMeta(vv.right))
      case _ if v.isInstanceOf[Or] =>
        val vv = v.asInstanceOf[Or]
        OrFilterNodeMeta("or", toFilterNodeMeta(vv.left), toFilterNodeMeta(vv.right))
      case _ if v.isInstanceOf[Expr] =>
        ExprFilterNodeMeta("expr", v.asInstanceOf[Expr])
    }
  }
}

case class AndOrOptMeta(__meta: MetaMeta, _clauses: List[FilterNodeMeta])


class AndOpt[T <: BaseNode](parent: T) extends BaseOpt {

  private var _isReady = false
  private[generator] val _clauses = ArrayBuffer[FilterNode]()


  def add(filterNode: FilterNode) = {
    _clauses += filterNode
    this
  }

  def end: T = {
    _isReady = true
    parent
  }

  override def toFilterNode: FilterNode = {
    val n = _clauses.reduce((a, b) => And(a, b))
    n
  }
}

class OrOpt[T <: BaseNode](parent: T) extends BaseOpt {

  private var _isReady = false
  private val _autogenTableName = UUID.randomUUID().toString.replaceAll("-", "")
  private var _tableName = _autogenTableName
  private[generator] var _clauses = ArrayBuffer[FilterNode]()


  def add(filterNode: FilterNode) = {
    _clauses += filterNode
    this
  }

  def tableName: String = {
    _tableName
  }

  def namedTableName(tableName: String): OrOpt[T] = {
    _tableName = tableName
    this
  }

  def desc(str: String): BaseNode = ???

  def end: T = {
    _isReady = true
    parent
  }

  override def toFilterNode: FilterNode = {
    val n = _clauses.reduce((a, b) => Or(a, b))
    n
  }
}

sealed abstract class UDFType {
  def sql: String
}

case object UDFUDFType extends UDFType {
  override def sql: String = "udf"
}

case object UDFUDAFType extends UDFType {
  override def sql: String = "udaf"
}

sealed abstract class LangType {
  def sql: String
}

case object LangScalaType extends LangType {
  override def sql: String = "scala"
}

case object LangJavaType extends LangType {
  override def sql: String = "java"
}

sealed abstract class VariableType {
  def sql: String
}

case object VariableSQLType extends VariableType {
  override def sql: String = "sql"
}

case object VariableShellType extends VariableType {
  override def sql: String = "shell"
}

case object VariableTextType extends VariableType {
  override def sql: String = "text"
}

case object VariableConfType extends VariableType {
  override def sql: String = "conf"
}

case object VariableDefaultParamType extends VariableType {
  override def sql: String = "defaultParam"
}

sealed abstract class VariableLifeTime {
  def sql: String
}

case object VariableSessionLifeTime extends VariableLifeTime {
  override def sql: String = "session"
}

case object VariableRequestLifeTime extends VariableLifeTime {
  override def sql: String = "request"
}

sealed abstract class VariableMode {
  def sql: String
}

case object VariableCompileMode extends VariableMode {
  override def sql: String = "compile"
}

case object VariableRuntimeMode extends VariableMode {
  override def sql: String = "Runtime"
}

File: python-client/tech/mlsql/byzer_client_sdk/python_lang/generator/builder.py
import dataclasses
import json
import uuid
from abc import ABCMeta, abstractmethod
from typing import Optional


class BaseNode(metaclass=ABCMeta):
    @abstractmethod
    def table_name(self) -> str:
        pass

    @abstractmethod
    def named_table_name(self, s: str) -> "BaseNode":
        pass

    @abstractmethod
    def tag(self, s: str) -> "BaseNode":
        pass

    @abstractmethod
    def end(self) -> "tech.mlsql.byzer_client_sdk.python_lang.generator.byzer.Byzer":
        pass

    @abstractmethod
    def options(self) -> "Options":
        pass

    @abstractmethod
    def to_block(self) -> str:
        pass

    @abstractmethod
    def get_tag(self) -> Optional[str]:
        pass

    @abstractmethod
    def to_json(self) -> str:
        pass

    @abstractmethod
    def from_json(self, s: str) -> "BaseNode":
        pass


class OptionValue(object):
    def __init__(self, value: str, quote_str: Optional[str]):
        self.value = value
        self.quote_str = quote_str


class Options(object):
    def __init__(self, parent: BaseNode):
        self.parent = parent
        self.o: dict[str, OptionValue] = {}

    def add(self, name: str, value: str) -> "Options":
        self.o[name] = OptionValue(value, None)
        return self

    def add_with_quoted_str(self, name: str, value: OptionValue) -> "Options":
        self.o[name] = value
        return self

    def end(self) -> BaseNode:
        return self.parent

    def items(self) -> dict[str, OptionValue]:
        return self.o

    def to_fragment(self) -> str:
        temp = []
        for k, v in self.o.items():
            if v.quote_str is None:
                if v.value is None:
                    temp.append(f"`{k}`=\"\"")
                else:
                    temp.append(f"`{k}`='''{v.value}'''")
            else:
                temp.append(f"`{k}`={v.quote_str}{v.value}{v.quote_str}")
        opts = " and ".join(temp)
        if len(self.o) != 0:
            return f"where {opts}"
        else:
            return ""


class MetaMeta(object):
    def __init__(self, name: str):
        self.name = name


class RawMeta:
    def __init__(self, __meta: MetaMeta, _tag: Optional[str], _is_ready: bool, _autogen_table_name: str, _code: str,
                 _table_name: str):
        self.__meta = __meta
        self._tag = _tag
        self._is_ready = _is_ready
        self._autogen_table_name = _autogen_table_name
        self._code = _code
        self._table_name = _table_name


@dataclasses.dataclass
class LoadMeta(object):
    _meta: MetaMeta
    _tag: Optional[str]
    _is_ready: bool
    _autogen_table_name: str
    _table_name: str
    _format: Optional[str]
    _path: Optional[str]
    _options: dict[str, OptionValue]


class Load(BaseNode):
    def __init__(self, parent: "tech.mlsql.byzer_client_sdk.python_lang.generator.byzer.Byzer"):
        self._is_ready = False
        self._autogen_table_name = str(uuid.uuid4())
        self._table_name = self._autogen_table_name
        self._format: Optional[str] = None
        self._path: Optional[str] = None
        self._options: Options = Options(self)
        self._tag: Optional[str] = None
        self._parent = parent

    def format(self, s: str) -> "Load":
        self._format = s
        return self

    def path(self, s: str) -> "Load":
        self._path = s
        return self

    def table_name(self) -> str:
        return self._table_name

    def named_table_name(self, s: str) -> BaseNode:
        self._table_name = s
        return self

    def tag(self, s: str) -> BaseNode:
        self._tag = s
        return self

    def end(self) -> "tech.mlsql.byzer_client_sdk.python_lang.generator.byzer.Byzer":
        self._is_ready = True
        return self._parent

    def options(self) -> Options:
        return self._options

    def to_block(self) -> str:
        return f"load {self._format}.`{self._path}` {self._options.to_fragment()} as {self._table_name};"

    def get_tag(self) -> Optional[str]:
        return self._tag

    def from_json(self, s: str) -> "BaseNode":
        v = json.loads(s)
        self._tag = v._tag
        self._is_ready = v._is_ready
        self._table_name = v._table_name
        self._format = v._format
        self._path = v._path
        self._options = Options(self)
        for key, value in self.options.items():
            self.options.add_with_quoted_str(key, value)
        return self

    def to_json(self) -> str:
        LoadMeta(
            _meta=MetaMeta(""),
            _tag=self._tag,
            _is_ready=self._is_ready,
            _autogen_table_name=self._autogen_table_name,
            _table_name=self._table_name,
            _format=self._format,
            _path=self._path,
            _options=self._options.items()
        )
        return self
